cranelift/interpreter/src/step.rs

```diff
@@ -199,45 +199,51 @@ where
         Ok(sum(imm, args)? as u64)
     };
 
+    // Interpret a unary instruction with the given `op`, assigning the resulting value to the
+    // instruction's results.
+    let unary = |op: fn(V) -> ValueResult<V>, arg: V| -> ValueResult<ControlFlow<V>> {
+        let ctrl_ty = inst_context.controlling_type().unwrap();
+        let res = unary_arith(arg, ctrl_ty, op, false)?;
+        Ok(assign(res))
+    };
+
     // Interpret a binary instruction with the given `op`, assigning the resulting value to the
     // instruction's results.
-    let binary = |op: fn(V, V) -> ValueResult<V>,
-                  left: V,
-                  right: V|
-     -> ValueResult<ControlFlow<V>> { Ok(assign(op(left, right)?)) };
+    let binary =
+        |op: fn(V, V) -> ValueResult<V>, left: V, right: V| -> ValueResult<ControlFlow<V>> {
+            let ctrl_ty = inst_context.controlling_type().unwrap();
+            let res = binary_arith(left, right, ctrl_ty, op, false)?;
+            Ok(assign(res))
+        };
 
     // Same as `binary_unsigned`, but converts the values to their unsigned form before the
     // operation and back to signed form afterwards. Since Cranelift types have no notion of
     // signedness, this enables operations that depend on sign.
     let binary_unsigned =
```

Comment:
Pre-existing: should this comment have said "Same as `binary`"?

Reply:
Yeah, I think so, I'll update this when doing the refactoring above ^